File: snippets/deno-rt2025/data-retention.ts
// @title: Data Retention Management
// @description: Implement custom data retention policies for device data
// @tags: data, retention, cleanup, management, storage

/*
 * Analysis Example
 * Custom Data Retention
 *
 * Get the list of devices, then go to each device removing the variables you chooses.
 *
 ** How to use:
 ** To analysis works, you need to add a new policy in your account. Steps to add a new policy:
 **  1 - Click the button "Add Policy" at this url: https://admin.tago.io/am;
 **  2 - In the Target selector, select the Analysis with the field set as "ID" and choose your Analysis in the list;
 **  3 - Click the "Click to add a new permission" element and select "Device" with the rule "Access" with the field as "Any";
 **  4 - To save your new Policy, click the save button in the bottom right corner;
 */

import { Analysis, Resources } from "jsr:@tago-io/sdk";
import dayjs from "npm:dayjs";

// The function startAnalysis will run when you execute your analysis
async function startAnalysis(context: any): Promise<void> {
  // Bellow is an empty filter.
  // Examples of filter:
  // { tags: [{ key: 'tag-key', value: 'tag-value' }]}
  // { name: 'name*' }
  // { name: '*name' }
  // { bucket: 'bucket-id' }
  const filter = {};

  const deviceQuery: any = {
    page: 1,
    fields: ["id"],
    filter,
    amount: 100,
  };

  const devices = await Resources.devices.list(deviceQuery);

  for (const deviceObj of devices) {
    const variables = ["variable1", "variable2"];
    const qty = 100; // remove 100 registers of each variable
    const end_date = dayjs().subtract(1, "month").toISOString(); // registers old than 1 month

    const removeOptions: any = { variables, qty, end_date };

    await Resources.devices
      .deleteDeviceData(deviceObj.id!, removeOptions)
      .then((result) => context.log(result))
      .catch((error) => context.log(error));
  }
}

Analysis.use(startAnalysis);

// To run analysis on your machine (external)
// Analysis.use(myAnalysis, { token: "YOUR-TOKEN" });

File: snippets/analysis/python-legacy/data-retention.py
# @title: Custom Data Retention
# @description: Automatically remove old data from devices based on custom retention policies
# @tags: data, retention, cleanup, management, automation

"""
Analysis Example
Custom Data Retention

Use your account token to get the list of devices, then go to each device removing the
variables you chooses.

Instructions
To run this analysis you need to add an account token to the environment variables,
To do that, go to your account settings, then token and copy your token.
Go the the analysis, then environment variables,
type account_token on key, and paste your token on value
"""

from tagoio_sdk import Analysis, Account, Device
from tagoio_sdk.modules.Utils.getTokenByName import getTokenByName


# The function myAnalysis will run when you execute your analysis
def my_analysis(context, scope: list):
    # reads the value of account_token from the environment variable
    account_token = next(
        (item for item in context.environment if item["key"] == "account_token"), None
    )

    if not account_token:
        raise ValueError("Missing 'account_token' in the environment variables")

    account = Account({"token": account_token["value"]})

    # Bellow is an empty filter.
    # Examples of filter:
    # { tags: [{ key: 'tag-key', value: 'tag-value' }]}
    # { name: 'name*' }
    # { name: '*name' }
    # { bucket: 'bucket-id' }
    filter = {}

    devices = account.devices.listDevice(
        {
            "page": 1,
            "fields": ["id"],
            "filter": filter,
            "amount": 100,
        }
    )

    for device_obj in devices:
        token = getTokenByName(account, device_obj["id"])
        device = Device({"token": token})

        variables = ["temperature"]
        qty = 100  # remove 100 registers of each variable
        end_date = "30 days"  # registers old than 30 days

        result = device.deleteData(
            {"variables": variables, "qty": qty, "end_date": end_date}
        )
        print(result)


# The analysis token in only necessary to run the analysis outside TagoIO
Analysis(params={"token": "MY-ANALYSIS-TOKEN-HERE"}).init(my_analysis)
